proposals/type-classes.md

```diff
@@ -237,60 +202,155 @@ fun addInts(a: Int, b: Int, with Monoid<Int>): Int = add(a, b)
+
+Orphan implementations are the reason that type classes have often been described as "anti-modular" as the most common way of dealing with them is through global coherency checks. This is necessary to ensure that two libraries have not defined incompatible implementations of a type class interface.
+
+Relaxing the orphan rules is a backwards-compatible change. If this proposal is accepted without permitting orphans it is useful to consider how they could be added in the future.
+
+Ideally we want to ban orphan implementations in libraries but not in executables; this allows a programmer to manually deal with coherence in their own code but prevents situations where adding a new library breaks code.
+
+### Package-based Approach to Orphans
+
+A simple way to allow orphan implementations is to replace the file-based restrictions with package-based restrictions. Because there are no restrictions on packages it is posible to do the following.
+
+```kotlin
+// In some library foo
+package foo.collections
+
+extension class Monoid<A> {
+   ...
 }
 ```
-Fails to compile lacking evidence that you can invoke `add(1,2)` since `add` is a polymorphic function that requires a `Monoid<Int>` inferred by `1` and `2` being of type `Int`.
-In such case the extension instance have to be explicitly defined.
+
 ```kotlin
-extension object AddingInts {
-  fun addInts(): Int = with(intext.IntMonoid) { add(1, 2) }
+// In some application that uses the foo library
+package foo.collections
+
+extension object : Monoid<Int> {
+   ...
 }
 ```
 
-Some of these examples where originally proposed by Roman Elizarov and the Kategory contributors where these features where originally discussed https://github.com/Kotlin/KEEP/pull/87
+This approach would not forbid orphan implementations in libraries but it would highly discourage them from providing them since it would involve writing code in the package namespace of another library.
+
+### Internal Modifier-based Approach to Orphans
+
+An alternate approach is to require that orphan implementations be marked `internal`. The full rules would be as follows:
+
+1. All orphan implementations must be marked `internal`
+2. All code which closes over an internal implementations must be marked internal. Code closes over a type class instance if it contains a static reference to such an implementation.
+3. Internal implementations defined in the same module are in scope for the current module.
+4. Internal implementations defined in other modules are not valid for type class resolution.
+
+This approach works well but it has a few problems.
+
+1. It forces applications that use orphan implementations to mark all their code as internal, which is a lot of syntactic noise.
```

Comment:
If I understood this correctly would this be anywhere else but the instance definition? Call sites would not have to be internal right?

Reply:
Call sites *do* have to be internal if they enclose over the type class, e.g. in `fun foo(i: Int, v: Int) = add(i, v)`, `foo` must be marked internal. This is necessary to stop the [set coherence problem](https://github.com/Kotlin/KEEP/pull/87#issuecomment-394488918). Otherwise one library could create a set (or other datatype) and then it can be consumed by another library with an incompatible type class instance.

Reply:
Type classes are first and foremost _interfaces_. Interfaces are useful because multiple libraries can program against them and then be used together. Library authors can program against them because they know that an implementing class has consistent behavior. Everything in this PR is designed to preserve that property.

Reply:
Let's address this thread here https://github.com/47deg/KEEP/pull/6#discussion_r198605608